Fix lambdas requiring extra line before End Of File

Scrawach opened the pull request:
fixes #96228

It seems that when parsing code [GDScriptTokinazer](https://github.com/godotengine/godot/blob/master/modules/gdscript/gdscript_tokenizer.cpp) was generating 2x`Token::DEDENT` at end instead of `Token::DEDENT` + `TOKEN::TK_EOF`. This was because lambdas have their own indent stack (`indent_stack_stack`) and it moved to default `indent_stack` after lambda parse, inside  `pop_expression_indented_block` method. As a result, this caused problems.

Meanwhile, `indent_stack` is clear in `check_indent()` method, when end of file reached:

https://github.com/godotengine/godot/blob/6395450b10d73bc3515763c7bbd1b2f5b7425d10/modules/gdscript/gdscript_tokenizer.cpp#L1173-L1178

I would have liked to reset `indent_stack_stack` here as well, but `pop_expression_indented_block` has error about empty stack.

https://github.com/godotengine/godot/blob/6395450b10d73bc3515763c7bbd1b2f5b7425d10/modules/gdscript/gdscript_tokenizer.cpp#L295-L299

So I had do it there.

Scrawach commented:
I added an extra commit by removing `clear` and replacing it with `pop_back`. This is because in the process I came across a bug in the solution, when another lambda is called inside one lambda and the bracket in the expression is not closed and not extra line before EOF:

```gdscript
func test(function1: Callable) -> void:
	function1.call()

func _ready():
	test(func(): test(func(): print("hello"))
```

If lambdas are called within each other, the unclosed parenthesis at the end of the expression, in addition to the expected `Expected closing ')' after calling arguments` error, will also report an empty lambda stack error from `pop_expression_indented_block`:

https://github.com/godotengine/godot/blob/6395450b10d73bc3515763c7bbd1b2f5b7425d10/modules/gdscript/gdscript_tokenizer.cpp#L295-L299

Chaosus commented:
Commits needs to be squashed (see https://docs.godotengine.org/en/latest/contributing/workflow/pr_workflow.html#modifying-a-pull-request)

Scrawach commented:
@Chaosus fixed.

But, PR don't pass static checks with file formatting. It require newline at end of file, but this issue (#96228) about miss newline at end of file. And that's we need for regression. What to do?

Chaosus commented:
> But, PR don't pass static checks with file formatting. It require newline at end of file, but this issue (https://github.com/godotengine/godot/issues/96228) about miss newline at end of file. And that's we need for regression. What to do?

If it's a test for error, you need to create it at `tests/scripts/runtime/errors` (look on other tests in that category as a reference).

Scrawach commented:
> If it's a test for error, you need to create it at tests/scripts/runtime/errors (look on other tests in that category as a reference).

No, this is not a error test. There shouldn't be any errors right now. But the style checker wants a empty line at end of the GDScript file. And if I add that line, the test is meaningless. Because it worked before with an empty line too. 

Issue #96228 about GDScript file without an empty line after lambda expression.

Chaosus commented:
I understood. Maybe the other developers knows how to fix it: cc @AThousandShips, cc @akien-mga. If it's impossible, then a test should probably just be removed.

akien-mga commented:
You need to add an exception for this file in the `modules/gdscript/tests/scripts/.editorconfig` file. I also suggest renaming the file as I put it in this diff to be a bit clearer / more concise:

```diff
diff --git a/modules/gdscript/tests/scripts/.editorconfig b/modules/gdscript/tests/scripts/.editorconfig
index 3a877f36f0..d4f466328f 100644
--- a/modules/gdscript/tests/scripts/.editorconfig
+++ b/modules/gdscript/tests/scripts/.editorconfig
@@ -1,6 +1,9 @@
 # Some tests handle invalid syntax deliberately; exclude relevant attributes.
 # See also the `file-format` section in `.pre-commit-config.yaml`.
 
+[parser/features/lambda_at_eof_without_newline.gd]
+insert_final_newline = false
+
 [parser/features/mixed_indentation_on_blank_lines.gd]
 trim_trailing_whitespace = false
 
```

Scrawach commented:
@akien-mga I did this, but Style Checker is still not happy. So additionally I add test script to exclusions for file-format in pre-commit config.

akien-mga reviewed (dismissed):
The test / dotfile changes look good.

I'm not an expert on the GDScript lambda implementation so can't say much about the actual fix, but it seems ok. I'll let GDScript maintainers confirm.

dalexeev commented on modules/gdscript/gdscript_tokenizer.cpp:
```diff
@@ -294,7 +294,11 @@ void GDScriptTokenizerText::push_expression_indented_block() {
 void GDScriptTokenizerText::pop_expression_indented_block() {
 	ERR_FAIL_COND(indent_stack_stack.is_empty());
-	indent_stack = indent_stack_stack.back()->get();
+
+	if (!_is_at_end()) {
+		indent_stack = indent_stack_stack.back()->get();
+	}
+
```
Probably the same change should be made to `GDScriptTokenizerBuffer`.

https://github.com/godotengine/godot/blob/89001f91d21ebd08b59756841426f540b154b97d/modules/gdscript/gdscript_tokenizer_buffer.cpp#L411-L415

Scrawach replied:
Yes, I think about that. But I was a little bit confused by the fact that `GDScriptTokenizerBuffer` has no `_is_at_end()` method. And right now I don't know how to write tests for this class. I'll look into it and add to PR if needed.

Scrawach replied:
@dalexeev so I added a `.bin.gd` test and also generated a `.gdc` file and separately tested the `GDScriptTokenizerBuffer` (via `GDScriptParser`). In summary, the error is not reproducible in `GDScriptTokenizerBuffer`. All is well.

Zireael07 commented:
Why close?

Scrawach commented:
@Zireael07 oops, it's my bad. I trying pull and merge new commits to my repository. A little confused)

Scrawach commented:
I changed the tests and I think it need re-review @akien-mga  

1. `test_gdscript.cpp`. I needed to run tests for the GDScriptTokenizerBuffer and use `.gdc` files to do so. However, the following code was problematic and caused parsing errors:

https://github.com/godotengine/godot/blob/0f95e9f8e665cedaa5aace5f6f86de2b4b5a8c60/modules/gdscript/tests/test_gdscript.cpp#L306-L310

`ERROR: Condition "total_len > 0" is true. Returning: ERR_INVALID_DATA`

All because of the addition of the last element. So I removed it. And it work fine. After that, added a branching and overloaded method for `test_parse` to start binary parsing.

2. `gdscript_test_runner.cpp`. I needed to generate a `.gdc` file for the tests, so I added a tool for that. I don't know if it will be needed. But I needed it :)

dalexeev commented on modules/gdscript/tests/scripts/parser/features/lambda_at_eof_without_newline.bin.gdc:
I don't think it's a good idea to include binary files in tests, because they are hard to review. As far as I know, the test system should automatically check binary mode, except for `*.textonly.gd`.

If you want to test `*.gdc` which can only be generated manually, I think it makes no sense, users don't create them manually.

Scrawach replied:
@dalexeev here I was referring to the code in `test_gdscript`, where we explicitly use `.gdc` files to test `Tokenizer Buffer`. 

https://github.com/godotengine/godot/blob/bdf625bd54958c737fa6b7213b07581cc91059ad/modules/gdscript/tests/test_gdscript.cpp#L328-L333

As I understand it, this code is used to run the tests “manually”. Easy and fast. And in the process of finding a bug, it was useful. 

Issue includes parser work, so I needed to test the parser in cooperation with `TokenizerBuffer`. That's why it appeared here. But I agree that binary files is hard to review. So I'm going to remove it from the PR. However, I am a bit confused by the fact that the code allows you to use `.gdc` files for testing, but they are not used anywhere in the repository)

Scrawach committed:
fix lambdas requiring extra line before End Of File

- ignore elements from lambda-specific indent stack in GDScriptTokenizer
- add test with this specific case for regression
- configure .editorconfig without insert final newline
- add test script file to exclude for file-format ci\cd
- fix incorrect read .gdc file in specific test
